server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java

```diff
@@ -270,4 +277,10 @@ protected Set<String> responseParams(RestApiVersion restApiVersion) {
         return responseParams();
     }
 
+    protected void setErrorTraceTransportHeader(RestRequest r) {
+        // set whether data nodes should send back stack trace based on the `error_trace` query parameter
+        if (threadContext != null) {
+            threadContext.putHeader("error_trace", r.param("error_trace", ERROR_TRACE_DEFAULT));
+        }
+    }
```

Comment:
There's a couple of problems with this approach, the thread context being mutable (the setter could be called multiple times in theory) and the instanceof check in ActionModule. It isn't the cleanest way. That said, before looking into alternatives, I'd look for core/infra review on this and get help there.